[TODO] Stop importing `pkg_resources`

It's deprecated and causes the following warning:
```console
E   DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
E   Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
```

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -32,9 +32,9 @@ requires-python = ">=3.7"
 dependencies = [
     "click",
     "click-default-group",
+    "importlib-resources>1.3; python_version<'3.9'",
     "incremental",
     "jinja2",
-    "setuptools",
     "tomli; python_version<'3.11'",
 ]
 

--- a/src/towncrier/_settings/load.py
+++ b/src/towncrier/_settings/load.py
@@ -112,6 +112,11 @@ def load_config_from_file(directory: str, config_file: str) -> Config:
     return parse_toml(directory, config)
 
 
+# Clean up possible temporary files on exit.
+_file_manager = ExitStack()
+atexit.register(_file_manager.close)
+
+
 def parse_toml(base_path: str, config: Mapping[str, Any]) -> Config:
     if "tool" not in config:
         raise ConfigError("No [tool.towncrier] section.", failing_option="all")
@@ -152,12 +157,10 @@ def parse_toml(base_path: str, config: Mapping[str, Any]) -> Config:
 
     template = config.get("template", _template_fname)
     if template.startswith("towncrier:"):
-        file_manager = ExitStack()
-        atexit.register(file_manager.close)
-
         resource_name = f"templates/{template.split('towncrier:', 1)[1]}.rst"
-        resource_path = resources.files("towncrier") / resource_name
-        file_manager.enter_context(resources.as_file(resource_path))
+        resource_path = _file_manager.enter_context(
+            resources.as_file(resources.files("towncrier") / resource_name)
+        )
 
         if not resource_path.is_file():
             raise ConfigError(

--- a/src/towncrier/_settings/load.py
+++ b/src/towncrier/_settings/load.py
@@ -3,10 +3,12 @@
 
 from __future__ import annotations
 
+import atexit
 import os
 import sys
 
 from collections import OrderedDict
+from contextlib import ExitStack
 from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Mapping
 
@@ -150,15 +152,20 @@ def parse_toml(base_path: str, config: Mapping[str, Any]) -> Config:
 
     template = config.get("template", _template_fname)
     if template.startswith("towncrier:"):
-        resource_name = "templates/" + template.split("towncrier:", 1)[1] + ".rst"
+        file_manager = ExitStack()
+        atexit.register(file_manager.close)
+
+        resource_name = f"templates/{template.split('towncrier:', 1)[1]}.rst"
         resource_path = resources.files("towncrier") / resource_name
-        if not resource_path.exists():
+        file_manager.enter_context(resources.as_file(resource_path))
+
+        if not resource_path.is_file():
             raise ConfigError(
                 "Towncrier does not have a template named '%s'."
                 % (template.split("towncrier:", 1)[1],)
             )
 
-        template = str(resource_path.resolve())
+        template = str(resource_path)
     else:
         template = os.path.join(base_path, template)
 

--- a/src/towncrier/_settings/load.py
+++ b/src/towncrier/_settings/load.py
@@ -10,8 +10,6 @@ from collections import OrderedDict
 from dataclasses import dataclass
 from typing import TYPE_CHECKING, Any, Mapping
 
-import pkg_resources
-
 from .._settings import fragment_types as ft
 
 
@@ -22,6 +20,12 @@ if TYPE_CHECKING:
         from typing_extensions import Literal
     else:
         from typing import Literal
+
+if sys.version_info < (3, 9):
+    import importlib_resources as resources
+else:
+    from importlib import resources
+
 
 if sys.version_info < (3, 11):
     import tomli as tomllib
@@ -147,13 +151,14 @@ def parse_toml(base_path: str, config: Mapping[str, Any]) -> Config:
     template = config.get("template", _template_fname)
     if template.startswith("towncrier:"):
         resource_name = "templates/" + template.split("towncrier:", 1)[1] + ".rst"
-        if not pkg_resources.resource_exists("towncrier", resource_name):
+        resource_path = resources.files("towncrier") / resource_name
+        if not resource_path.exists():
             raise ConfigError(
                 "Towncrier does not have a template named '%s'."
                 % (template.split("towncrier:", 1)[1],)
             )
 
-        template = pkg_resources.resource_filename("towncrier", resource_name)
+        template = str(resource_path.resolve())
     else:
         template = os.path.join(base_path, template)
 

--- a/src/towncrier/newsfragments/496.feature
+++ b/src/towncrier/newsfragments/496.feature
@@ -1,0 +1,1 @@
+Towncrier no longer depends on setuptools & uses importlib.resources (or its backport) instead.

--- a/src/towncrier/test/helpers.py
+++ b/src/towncrier/test/helpers.py
@@ -32,7 +32,7 @@ def read_pkg_resource(path: str) -> str:
     """
     Read *path* from the towncrier package.
     """
-    return (resources.files("towncrier") / path).read_text()
+    return (resources.files("towncrier") / path).read_text("utf8")
 
 
 def with_isolated_runner(fn: Callable[..., Any]) -> Callable[..., Any]:

--- a/src/towncrier/test/helpers.py
+++ b/src/towncrier/test/helpers.py
@@ -1,10 +1,18 @@
 from __future__ import annotations
 
+import sys
+
 from functools import wraps
 from pathlib import Path
 from typing import Any, Callable
 
 from click.testing import CliRunner
+
+
+if sys.version_info < (3, 9):
+    import importlib_resources as resources
+else:
+    from importlib import resources
 
 
 def read(filename: str | Path) -> str:
@@ -18,6 +26,13 @@ def write(path: str | Path, contents: str) -> None:
     p = Path(path)
     p.parent.mkdir(parents=True, exist_ok=True)
     p.write_text(contents)
+
+
+def read_pkg_resource(path: str) -> str:
+    """
+    Read *path* from the towncrier package.
+    """
+    return (resources.files("towncrier") / path).read_text()
 
 
 def with_isolated_runner(fn: Callable[..., Any]) -> Callable[..., Any]:

--- a/src/towncrier/test/test_format.py
+++ b/src/towncrier/test/test_format.py
@@ -4,11 +4,10 @@
 
 from collections import OrderedDict
 
-import pkg_resources
-
 from twisted.trial.unittest import TestCase
 
 from .._builder import render_fragments, split_fragments
+from .helpers import read_pkg_resource
 
 
 class FormatterTests(TestCase):
@@ -119,9 +118,7 @@ Bugfixes
 - Web fixed. (#3)
 """
 
-        template = pkg_resources.resource_string(
-            "towncrier", "templates/default.rst"
-        ).decode("utf8")
+        template = read_pkg_resource("templates/default.rst")
 
         fragments = split_fragments(fragments, definitions)
         output = render_fragments(
@@ -208,9 +205,7 @@ Misc
 - xxbar, xx1, xx9, xx142
 """
 
-        template = pkg_resources.resource_string(
-            "towncrier", "templates/default.rst"
-        ).decode("utf8")
+        template = read_pkg_resource("templates/default.rst")
 
         fragments = split_fragments(fragments, definitions)
         output = render_fragments(
@@ -266,9 +261,7 @@ Features
   a a (#3)
 """
 
-        template = pkg_resources.resource_string(
-            "towncrier", "templates/default.rst"
-        ).decode("utf8")
+        template = read_pkg_resource("templates/default.rst")
 
         fragments = split_fragments(fragments, definitions)
         output = render_fragments(
@@ -317,9 +310,7 @@ Features
 - a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a a (#3)
 """  # NOQA
 
-        template = pkg_resources.resource_string(
-            "towncrier", "templates/default.rst"
-        ).decode("utf8")
+        template = read_pkg_resource("templates/default.rst")
 
         fragments = split_fragments(fragments, definitions)
         output = render_fragments(

--- a/src/towncrier/test/test_write.py
+++ b/src/towncrier/test/test_write.py
@@ -7,14 +7,13 @@ from collections import OrderedDict
 from pathlib import Path
 from textwrap import dedent
 
-import pkg_resources
-
 from click.testing import CliRunner
 from twisted.trial.unittest import TestCase
 
 from .._builder import render_fragments, split_fragments
 from .._writer import append_to_newsfile
 from ..build import _main
+from .helpers import read_pkg_resource
 
 
 class WritingTests(TestCase):
@@ -92,9 +91,7 @@ Old text.
 
         fragments = split_fragments(fragments, definitions)
 
-        template = pkg_resources.resource_string(
-            "towncrier", "templates/default.rst"
-        ).decode("utf8")
+        template = read_pkg_resource("templates/default.rst")
 
         append_to_newsfile(
             tempdir,
@@ -205,9 +202,7 @@ Old text.
 
         fragments = split_fragments(fragments, definitions)
 
-        template = pkg_resources.resource_string(
-            "towncrier", "templates/default.rst"
-        ).decode("utf8")
+        template = read_pkg_resource("templates/default.rst")
 
         append_to_newsfile(
             tempdir,
@@ -242,9 +237,7 @@ Old text.
         definitions = {}
         fragments = split_fragments(fragments={}, definitions=definitions)
 
-        template = pkg_resources.resource_string(
-            "towncrier", "templates/default.rst"
-        ).decode("utf8")
+        template = read_pkg_resource("templates/default.rst")
 
         content = render_fragments(
             template=template,

